### PR TITLE
parallel dls

### PR DESCRIPTION
**Pull request by rawdaGastan**
- https://github.com/codescalers/cloud4students/issues/112

### CODE REVIEW

File: server/models/api_inputs.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
+// Package models for database models
+package models
+
```
Comment (xmonader) [inline server/models/api_inputs.go:3]: I believe these api inputs should be maintained on the routes package?

File: server/routes/server.go [hunk 2]
```diff
@@ -111,6 +117,13 @@ func NewServer(file string) (server *Server, err error) {
 
 	// notify admins
 	go router.NotifyAdmins()
```
Comment (xmonader) [inline server/routes/server.go]: why not send the ctx as well to the NotifyAdmins go routine?

File: server/routes/server.go [hunk 3]
```diff
@@ -111,6 +117,13 @@ func NewServer(file string) (server *Server, err error) {
...
+	// periodic deployments
```
Comment (xmonader) [inline server/routes/server.go]: I believe we need to refactor a little bit, routes is more than routes now,

I suggest:
- App struct (that has the logger, router, .. etc)
- Also it should have a method to start background workers e.g mailsender, periodic checks .. etc

File: server/routes/server.go [hunk 4]
```diff
@@ -111,6 +117,13 @@ func NewServer(file string) (server *Server, err error) {
...
+	go router.periodicRequests(ctx)
+	go router.periodicDeploy(ctx)
+
+	// check pending deployments
```
Comment (xmonader) [inline server/routes/server.go]: the following calls aren't clear to me from the first sight.

File: server/streams/types.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+	// ReqVMStreamName stream name
+	ReqVMStreamName = "vms-req"
+	// ReqK8sStreamName stream name
+	ReqK8sStreamName = "k8s-req"
+)
+
+// ErrResponse is a response for deployment requests
+type ErrResponse struct {
+	Code *int
```
Comment (xmonader) [inline server/streams/types.go:37]: Is this code, a substrate error code that could be missing?

File: server/streams/types.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+	"github.com/threefoldtech/tfgrid-sdk-go/grid-client/workloads"
+)
+
+const (
+	// DeployVMConsumerGroupName consumer group name
+	DeployVMConsumerGroupName = "vms-group"
+	// DeployK8sConsumerGroupName consumer group name
+	DeployK8sConsumerGroupName = "k8s-group"
+	// DeployNetConsumerGroupName consumer group name
+	DeployNetConsumerGroupName = "nets-group"
```
Comment (xmonader) [inline server/streams/types.go:15]: Not sure I'm following with this one

File: server/routes/k8s_router.go [hunk 7]
```diff
@@ -42,101 +33,44 @@ func (r *Router) K8sDeployHandler(w http.ResponseWriter, req *http.Request) {
...
-		return
-	}
-	if err != nil {
-		log.Error().Err(err).Send()
-		writeErrResponse(req, w, http.StatusInternalServerError, internalServerErrorMsg)
-		return
+	for resCode == nil {
+		r.mutex.Lock()
+		resCode = r.k8sRequestResponse[key].Code
+		r.mutex.Unlock()
```
Comment (AbdelrahmanElawady) [inline server/routes/k8s_router.go]: if I understand correctly this runs a loop waiting for response which is not efficient as it keeps locking and unlocking the mutex, using channel to communicate or `sync.Cond` might be better.

File: server/routes/vm_router.go [hunk 8]
```diff
@@ -2,137 +2,73 @@
...
-		PublicIP:          vm.ComputedIP,
-		SRU:               diskSize,
-		CRU:               uint64(vm.CPU),
-		MRU:               uint64(vm.Memory),
-		ContractID:        contractID,
-		NetworkContractID: networkContractID,
+	for resCode == nil {
+		r.mutex.Lock()
+		resCode = r.vmRequestResponse[key].Code
+		r.mutex.Unlock()
```
Comment (AbdelrahmanElawady) [inline server/routes/vm_router.go]: same as before

File: server/routes/consumer.go [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+	if err != nil {
+		if errors.Is(err, redis.Nil) {
+			return vms, nil
+		}
+		return vms, errors.Wrap(err, "failed to read vm stream deployment")
+	}
+
+	for _, s := range result {
+		for i, message := range s.Messages {
+			// consume 5 deployments only
```
Comment (xmonader) [inline server/routes/consumer.go]: Can't that info be encoded in the XREAD itself with the COUNT option?

File: server/routes/deployer_client.go [hunk 10]
```diff
@@ -163,26 +176,32 @@ func (r *Router) deployVM(ctx context.Context, vmInput DeployVMInput, sshKey str
...
+	err = r.redis.PushVM(streams.NetDeployment{DL: &network}, streams.VMDeployment{DL: &dl})
 	if err != nil {
 		return nil, 0, 0, 0, err
 	}
 
-	// deploy vm
-	dl := workloads.NewDeployment(vmInput.Name, nodeID, "", nil, network.Name, []workloads.Disk{disk}, nil, []workloads.VM{vm}, nil)
-	err = r.tfPluginClient.DeploymentDeployer.Deploy(ctx, &dl)
+	// wait for deployments
+	for !r.vmDeployed {
```
Comment (AbdelrahmanElawady) [inline server/routes/deployer_client.go]: a channel would be better here

File: server/routes/deployer_client.go [hunk 11]
```diff
@@ -61,20 +62,32 @@ func (r *Router) deployK8sClusterWithNetwork(k8sDeployInput K8sDeployInput, sshK
...
+	// add network and cluster to be deployed
+	err = r.redis.PushK8s(streams.NetDeployment{DL: &network}, streams.K8sDeployment{DL: &cluster})
 	if err != nil {
-		return 0, 0, 0, errors.Wrapf(err, "failed to deploy network on nodes %v", network.Nodes)
+		return 0, 0, 0, err
+	}
+
+	// wait for deployments
+	for !r.k8sDeployed {
+		continue
```
Comment (AbdelrahmanElawady) [inline server/routes/deployer_client.go]: same here

File: server/routes/deployer_client.go [hunk 12]
```diff
@@ -395,13 +414,57 @@ func buildK8sCluster(node uint32, sshKey, network string, k K8sDeployInput) (wor
...
 
-// generate random names for network
-func generateNetworkName() string {
-	var seededRand *rand.Rand = rand.New(rand.NewSource(time.Now().UnixNano()))
-	charset := "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
-	name := make([]byte, 4)
-	for i := range name {
-		name[i] = charset[seededRand.Intn(len(charset))]
+func (r *Router) periodicRequests(ctx context.Context) {
+	ticker := time.NewTicker(time.Second * 6)
```
Comment (AbdelrahmanElawady) [inline server/routes/deployer_client.go]: 6 secs should be parametrized?

File: server/routes/deployer_client.go [hunk 13]
```diff
@@ -395,13 +414,57 @@ func buildK8sCluster(node uint32, sshKey, network string, k K8sDeployInput) (wor
...
+				log.Error().Err(err).Msg("failed to batch deploy network")
+			}
+		}
+
+		if len(vms) > 0 {
+			err := r.tfPluginClient.DeploymentDeployer.BatchDeploy(ctx, vms)
+			if err != nil {
+				log.Error().Err(err).Msg("failed to batch deploy vm")
+			}
+			r.vmDeployed = true
```
Comment (AbdelrahmanElawady) [inline server/routes/deployer_client.go]: these bools are accessed from different goroutines they should be locked?

**Review by AbdelrahmanElawady — COMMENTED** [review]
some variables are accessed in different goroutines `-race` flag when using go run or go test should show some data races. Also I think it would be better to decouple the deployment process from returning deployments but thats maybe in another pr.
but overall great work.

File: server/models/database.go [hunk 14]
```diff
@@ -326,3 +328,22 @@ func (d *DB) GetMaintenance() (Maintenance, error) {
...
+// ListNotifications returns a list of notifications for a user.
+func (d *DB) ListNotifications(userID string) ([]Notification, error) {
+	var res []Notification
+	query := d.db.Where("user_id = ?", userID).Find(&res)
+	return res, query.Error
+}
+
+// UpdateNotification updates seen field for notification
+func (d *DB) UpdateNotification(id int, seen bool) error {
+	return d.db.Model(&Maintenance{}).Where("id = ?", id).Updates(map[string]interface{}{"seen": seen}).Error
```
Comment (xmonader) [inline server/models/database.go]: shouldn't that be Notification instead ?

File: server/routes/notification_router.go [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+	}
+
+	err = r.db.UpdateNotification(id, true)
+	if err != nil {
+		log.Error().Err(err).Send()
+		writeErrResponse(req, w, http.StatusInternalServerError, internalServerErrorMsg)
+		return
+	}
+
+	writeMsgResponse(req, w, "Notifications is updated", "")
```
Comment (xmonader) [inline server/routes/notification_router.go]: is -> are

**Review by xmonader — COMMENTED** [review]
Please refactor the routes to have an App struct for dependency instead of the route

```
App {
 router
 db
 redis
 mutexes
 loggers. .. 
}
```

**Review by xmonader — APPROVED** [review]

**Pull request merged**